src/modules/import.js

```diff
@@ -128,237 +71,226 @@ export const closeImport = () => ({
+    };
+  }
+  return state;
 };
 
-/**
- * Returns the state after the import failed action.
- *
- * @param {Object} state - The state.
- * @param {Object} action - The action.
- *
- * @returns {Object} The new state.
- */
-export const doImportFailed = (state, action) => ({
-  ...state,
-  error: action.error,
-  progress: 100,
-  status: PROCESS_STATUS.FAILED
-});
-
-/**
- * Returns the state after the file type selected action.
- *
- * @param {Object} state - The state.
- * @param {Object} action - The action.
- *
- * @returns {Object} The new state.
- */
-export const doImportFileTypeSelected = (state, action) => ({
-  ...state,
-  fileType: action.fileType
-});
-
-/**
- * Returns the state after the file name selected action.
- *
- * @param {Object} state - The state.
- * @param {Object} action - The action.
- *
- * @returns {Object} The new state.
- */
-export const doImportFileNameSelected = (state, action) => ({
-  ...state,
-  fileName: action.fileName
-});
+export const startImport = () => {
+  return (dispatch, getState) => {
+    const state = getState();
+    const { ns, dataService: { dataService }, importData } = state;
+    const { fileName, fileType, fileStats: { size } } = importData;
+    const source = fs.createReadStream(fileName, 'utf8');
 
-/**
- * Returns the state after the open action.
- *
- * @returns {Object} The new state.
- */
-export const doOpenImport = () => ({
-  ...INITIAL_STATE,
-  isOpen: true
-});
+    const progress = createProgressStream({
+      length: size,
+      time: 250 /* ms */
+    });
 
-/**
- * Returns the state after the close action.
- *
- * @param {Object} state - The state.
- *
- * @returns {Object} The new state.
- */
-export const doCloseImport = (state) => ({
-  ...state,
-  isOpen: false
-});
+    progress.on('progress', function(info) {
+      console.log(info);
+      dispatch(updateProgress(info.percentage));
+
+      /*
+      {
+        percentage: 9.05,
+        transferred: 949624,
+        length: 10485760,
+        remaining: 9536136,
+        eta: 42,
+        runtime: 3,
+        delta: 295396,
+        speed: 949624
+      }
+      */
+    });
+    
+    // const f = throttle(function() {
+    //   debug('progress', (source.bytesRead / size) * 100);
+    //   dispatch(updateProgress((source.bytesRead / size) * 100));
+    // }, 250);
+
+    // const progress = new stream.Transform({
+    //   objectMode: true,
+    //   transform(chunk, encoding, callback) {
+    //     f();
+    //     callback(null, chunk);
+    //   }
+    // });
+
+    const deserializer = createEJSONDeserializer();
+
+    let parser;
+    if (fileType === 'csv') {
+      parser = createCSVParser();
+    } else {
+      parser = createJSONParser();
+    }
+
+    const dest = createCollectionWriteStream(dataService, ns);
+    debug('executing pipeline');
+
+    dispatch(importStarted(source, dest));
+    stream.pipeline(source, parser, deserializer, progress, dest, function(err, res) {
```

Comment:
i am very happy we can now just do `stream.pipeline()`, pump was good, but it's soooo nice to do this natively!!

Reply:
Me too!